### PR TITLE
feat(ccls-1788): implementation of notifications search

### PR DESCRIPTION
**Pull request by keith-h**

### CODE REVIEW

File: src/integrationTest/java/uk/gov/laa/ccms/caab/client/EbsApiClientIntegrationTest.java [hunk 1]
```diff
@@ -154,6 +158,40 @@ public void testGetAmendmentTypes_returnData() throws Exception {
...
+
+    assertNotNull(result);
+    assertEquals(userDetailsJson, objectMapper.writeValueAsString(result));
+
+  }
+  @Test
+  public void testGetUsers_notFound() {
+    Integer providerId = 123;
+    String expectedMessage = String.format(USERS_ERROR_MESSAGE, providerId);
+    wiremock.stubFor(get(String.format("/users/provider-id=%s", providerId))
```
Comment (PhilDigitalJustice) [inline src/integrationTest/java/uk/gov/laa/ccms/caab/client/EbsApiClientIntegrationTest.java:184]: Is this the correctly mocked endpoint?

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/notification/NotificationSearchValidator.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      if (!clientSurname.matches(FIRST_CHARACTER_MUST_BE_ALPHA)) {
+        errors.rejectValue("clientSurname", "invalid.surname",
+            "Your input for 'Client surname' is invalid. "
+                + "The first character must be a letter. Please amend your entry.");
+      } else if (!clientSurname.matches(CHARACTER_SET_C)) {
+        errors.rejectValue("clientSurname", "invalid.surname-char",
+            "Your input for 'Client surname' contains an invalid character. "
+                + "Please amend your entry.");
+      } else if (clientSurname.matches(DOUBLE_SPACE)) {
+        errors.rejectValue("caseReference", "invalid.case-ref",
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/notification/NotificationSearchValidator.java:161]: DO you mean caseReference here?

File: src/integrationTest/java/uk/gov/laa/ccms/caab/client/EbsApiClientIntegrationTest.java (hunk 1 above)
Comment (keith-h) [reply src/integrationTest/java/uk/gov/laa/ccms/caab/client/EbsApiClientIntegrationTest.java:184]: er.  no.  Will update

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/notification/NotificationSearchValidator.java (hunk 2 above)
Comment (keith-h) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/notification/NotificationSearchValidator.java:161]: I do indeed.  Will amend.

File: src/main/java/uk/gov/laa/ccms/caab/controller/notifications/ActionsAndNotificationsController.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+      // do not set a notification type and reset the search criteria
+      resetCriteria(criteria);
+    }
+    // Get the sort parameters
+    if (StringUtils.isNotEmpty(sort)) {
+      criteria.setSort(sort);
+    }
+  }
+
+  private static void resetCriteria(NotificationSearchCriteria criteria) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/controller/notifications/ActionsAndNotificationsController.java:191]: can this be moved to the NotificationSearchCriteria as a constructor?

Comment (keith-h) [reply src/main/java/uk/gov/laa/ccms/caab/controller/notifications/ActionsAndNotificationsController.java:191]: yes.  Will create new object.

File: src/main/resources/templates/partials/header.html [hunk 4]
```diff
@@ -28,7 +28,7 @@
                         </a>
                     </li>
                     <li class="govuk-header__navigation-item" th:classappend="${pageCategory == 'actions'} ? 'govuk-header__navigation-item--active'">
-                        <a class="govuk-header__link" href="/notifications">
+                        <a class="govuk-header__link" href="/notifications?notification_type=all">
```
Comment (PhilDigitalJustice) [inline src/main/resources/templates/partials/header.html:31]: could 'all' be the default set by the controller, then no need to specify it here?

**Review by PhilDigitalJustice — CHANGES_REQUESTED** [review]
A few comments please amend

**Review by PhilDigitalJustice — APPROVED** [review]
LGTM

**Pull request merged**